rfcs/text/0000-rfc-application-layer-bundle-types-and-definitions.md

```diff
@@ -0,0 +1,431 @@
+
+
+1. Allow for one level of subtyping to refine the types.
+
+
+1. Further refine the type by referencing the process and protocol, used to create the bundle and other type specific information as required.
+
+
+1. Add a type.json file to each bundle to specify the bundle type information.
+
+
+1. Specify the desired types in JSON schema.
+
+
+1. Document bundle types and shapes in the data portal.
+
+#### Why JSON and JSON Schema instead of media type syntax?
+JSON is more human readable, and can express more complex types and can be validated. Also we are significantly invested in JSON schema and have the tools and processes to support the schema lifecycle in place.
+
+
+### Core Types
+
+![Core Types](../images/0000-rfc-application-layer-bundle-types-coretypes.png)
+
+#### Assay Bundle
+
+Umbrella type for Assays/Measurements with the following subclasses:
+
+
+1. “Sequencing Assay” - representing the current sequencing ingest bundles.
+
+1. “Imaging Assay” - representing the current imaging ingest bundles. 
+
+1. Other high level assay types.
+
+
+> _Note: It may take some discussion to decide on an effective set of subclasses. Sequencing assay may need to change to “2nd Gen Sequencing Assay” or “NGS Sequencing Assay” for example._
+
+#### Analysis Bundle
+Umbrella type used to represent an analysis with the following possible subclasses:
+
+
+1. “Alignment-Expression” - representing our current secondary analysis pipelines.
+
+
+1. “Expression Matrix” - a possible subtype representing the expression matrix generation.
+
+
+1. “Expression Normalization” - a possible subtype representing the expression matrix normalization output for example.
+
+
+1. “Clustering” - a possible subtype representing cell clustering analyses output for example.
+
+> _Note: The exact names of the subclasses above can likely be improved upon. Please suggest._
+
+#### Reference Bundle
+Umbrella type representing a reference file used in analysis. Subclasses TBD.
+
+#### Entity Bundle
+A container to hold an explicit view of all of the metadata associated with a given domain object such as project or donor. Possible subclasses:
+
+1. Project  - currently one must iterate over all bundles in a project to build the full project graph. For example, a “Project Entity” bundle could hold the entire metadata graph for a project version in both json and tsv formats.
+
+1. Donor - similarly, a donor entity bundle could hold the entire metadata graph for a given donor in multiple formats.
+
+#### Event Bundle 
+
+Representing state change of a set of bundles relating to a domain entity or an arbitrary set of bundles. This could be used for example to implement a “data group” indicating that all of the Assay Bundles corresponding to a specific library prep instance have been ingested and secondary analysis on this “version complete” data set can commence. 
+
+### Example JSON Files
+
+#### Assay Bundle
```

Comment:
Merging links.json, which is part of the experimental metadata with bundle types defeats the goal of the metadata group at separating the metadata layer from the bundle layer.

Reply:
The goal here is to tag the bundles with the process and protocol that created them in a uniform manner without disrupting the current system. Links.json use different file names for analysis vs assay bundles for the process and protocol links. Here the process and protocol are normalized across bundle types. 

Also by using the protocol and process id this gives the ultimate in type refinement on what is in this bundle. 

I do not see this as breaking a separation of concerns. The ids of the process and protocol used to create the bundle should not be uniquely visible to the "experimental metadata".  Our "processing model" as well should be able to benefit from the terms created in the experimental model to the extent that it is useful.

Reply:
It the processes and protocols being referenced are metadata instances, this is very confusing

Reply:
> Yes `provenance.document_id` is the type.json file DSS UUID. It's not clear to me exactly what the `provenance.submission_date` means in this context and if this maps reliably to the DSS file version in general.
> 
> I have updated the doc to reflect the above.

Provenance section is added by ingest.  The submission_date may correspond to the ingest, however, is in a slightly different format so it should not be counted on. This is problematic and there is a ticket about this weirdness.

Reply:
Also, this does not couple the experimental metadata with the bundle metadata, it couples the bundle metadata with the experimental metadata in order to define the bundle type in terms of the experimental metadata. This is a use, not a pollution of the experimental metadata.

Reply:
We can easily add new data to links.json too to address these issues.